### PR TITLE
Eng 11743 stats no reset

### PR DESCRIPTION
**Pull request by xinjiacs**
Keep UAC stats after catalog update.

### CODE REVIEW

File: src/frontend/org/voltdb/ProcedureStatsCollector.java [hunk 1]
```diff
@@ -133,14 +133,14 @@
...
+        m_procName = catProc.getClassname();
     }
 
     /**
      * Called when a procedure begins executing. Caches the time the procedure starts.
      */
-    public final void beginProcedure() {
-        if (m_invocations % timeCollectionInterval == 0) {
+    public final void beginProcedure(boolean isSystemProc) {
+        if (m_invocations % timeCollectionInterval == 0 || (isSystemProc && isProcedureUAC())) {
```
Comment (luyangco) [inline src/frontend/org/voltdb/ProcedureStatsCollector.java]: `isProcedureUAC()` has a string comparison so still some performance impact on transactional path if system procedures run frequently.

File: src/frontend/org/voltdb/StatsAgent.java [hunk 2]
```diff
@@ -572,26 +610,39 @@ public VoltTable getStatsAggregate(
...
     {
         assert selector != null;
-        final NonBlockingHashMap<Long, NonBlockingHashSet<StatsSource>> siteIdToStatsSources = registeredStatsSources.get(selector);
+        NonBlockingHashMap<Long, NonBlockingHashSet<StatsSource>> siteIdToStatsSources =
+                m_registeredStatsSources.get(selector);
 
         // There are cases early in rejoin where we can get polled before the server is ready to provide
         // stats.  Just return null for now, which will result in no tables from this node.
-        if (siteIdToStatsSources == null || siteIdToStatsSources.isEmpty()) {
+        if (selector == StatsSelector.PROCEDURE) {
```
Comment (luyangco) [inline src/frontend/org/voltdb/StatsAgent.java]: Please comment on why `StatsSelector.PROCEDURE` needs separate handling

Comment (luyangco) [inline src/frontend/org/voltdb/StatsAgent.java]: Also the comments above should move down the line 640.

File: src/frontend/org/voltdb/ProcedureStatsCollector.java (hunk 1 above)
Comment (xinjiacs) [inline src/frontend/org/voltdb/ProcedureStatsCollector.java]: I put "isSystemProc" first as boolean operation hoping the early return helps. 
So for non-system procedure call, it should not affect the performance. For system procedure call, string prefix check may affect. Let's check performance benchmark jenkins first.

If bad on AdHoc (frequently called), we can skip prefix string "org.voltdb.sysprocs.". It should be sufficient enough to check a couple characters.

File: src/frontend/org/voltdb/ProcedureStatsCollector.java [hunk 3]
```diff
@@ -163,7 +163,7 @@ public final void endProcedure(
             {
                 if (Math.abs(delta) > 1000000000)
                 {
-                    log.info("Procedure: " + m_catProc.getTypeName() +
+                    log.info("Procedure: " + m_procName +
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/ProcedureStatsCollector.java]: getTypeName() is null, and this function is a generic function from catalog item. 
getClassname() has real meaning, specifically for PROCEDURE.

File: src/frontend/org/voltdb/ProcedureStatsCollector.java [hunk 4]
```diff
@@ -377,4 +377,43 @@ public long getInvocations() {
...
+
+    @Override
+    public int hashCode() {
+        return super.hashCode() + m_partitionId + m_procName.hashCode();
+    }
+
+    /**
+     * @return true if this procedure statistics should be reset
+     */
+    public boolean resetAfterCatalogChange() {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/ProcedureStatsCollector.java]: I did it on purpose with TODO next if we can allow more procedures to be kept.

File: src/frontend/org/voltdb/ProcedureRunner.java [hunk 5]
```diff
@@ -273,7 +274,7 @@ public ClientResponseImpl call(Object... paramListIn) {
         assert(m_batch.size() == 0);
 
         try {
-            m_statsCollector.beginProcedure();
+            m_statsCollector.beginProcedure(isSystemProcedure());
```
Comment (yzhang1991) [inline src/frontend/org/voltdb/ProcedureRunner.java]: I guess this `isSystemProcedure` can be stored as a field of statsCollector?

**Pull request merged**